## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [5]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [14]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#utilize data frame that only shows quantity of order based on Category &
#Quantity purchased from the retained data
order_streamline = df.loc()[(df["category"] == "software") & (df["qty"] > 2), :]
order_streamline.head(20)

first       last                                  job  \
2        Shannon     Watson                         Immunologist   
10          Kyle   Williams  Engineer, broadcasting (operations)   
30       Deborah       Cook            Radiographer, therapeutic   
33       Cynthia     Palmer             Surveyor, rural practice   
37       Bradley    Spencer                    Warehouse manager   
39       Mallory   Mitchell                       Quarry manager   
46        Ashley  Hernandez                         Retail buyer   
62         James    Carlson                            Press sub   
90         Susan      Lewis                    Therapist, sports   
98      Nicholas   Williams                       Radio producer   
106     Clarence      Chang                      Records manager   
120      Brandon      Allen                  Geologist, wellsite   
121       Thomas    Wiggins                            Paramedic   
125       Claire    Serrano                 Biomedical scientist   
127         Mary    Ramirez             Customer service manager   
131         Mark     Turner              Education administrator   
135  Christopher  Rodriguez                    Buyer, industrial   
138        David       Hart           Designer, interior/spatial   
139       Robert     Taylor                      Learning mentor   
141      Jessica   Gonzalez                  Editorial assistant   

                      phone                                     email  \
2         687.737.9424x8503                     swatson8146@payne.net   
10          +1-432-234-2942  communitywilliams9024@taylor-padilla.biz   
30       747.841.2427x57672                   cook.d@martin-yoder.com   
33      +1-843-385-5943x730                 palmer.cynthia@weaver.com   
37      +1-260-443-8861x751                   b.spencer.1089@ross.com   
39        335-735-8539x4710  mitchell.mallory.7004@powell-english.com   
46             264-496-5412      ashley.later.185@hernandez-shaw.info   
62         001-347-367-6286            jcarlson936@rhodes-mcguire.biz   
90               2913618427              s.lewis.5434@parker-reid.biz   
98        393.560.7337x2039         nicholas.williams.6317@strong.biz   
106        001-439-537-5844          chang.c.9094@gardner-sawyer.info   
120            705-244-4626                    ballen2525@houston.net   
121        380-613-5572x875                         twiggins@ball.com   
125    001-667-417-7335x439             claireserrano6407@rivera.info   
127            893-575-5590                    ramirez.mary@diaz.info   
131           (640)604-9066                 markturner4518@garcia.com   
135        001-723-752-6943           rodriguezchristopher@taylor.com   
138              2229562101                  hart.david.9828@wood.com   
139       743-370-7870x7946             rtaylor7129@navarro-smith.com   
141  001-766-709-4338x01962                 gonzalezj@horn-fields.com   

     client_id  order_id  order_date  order_week  order_year         item_id  \
2        57113   1902144  2023-01-29           4        2023  ABE59463-05-7E   
10       81012   5775924  2023-04-12          15        2023  DCD61783-36-6B   
30       10351   3052159  2023-01-27           4        2023  MAU13971-98-9B   
33       95847   6155436  2023-03-15          11        2023  CCA32097-88-6M   
37       48015   5300796  2023-04-18          16        2023  AEM31850-24-5M   
39       67516   7012133  2023-04-24          17        2023  XAU13501-34-9A   
46       61647   4950072  2023-01-30           5        2023  AEE37351-26-7B   
62       28011   8527601  2023-05-21          20        2023  XAU13501-34-9A   
90       13436   8468237  2023-02-06           6        2023  EBA49136-27-6B   
98       56590   7777768  2023-01-15           2        2023  UZX16759-11-7E   
106      57762   3048979  2023-05-05          18        2023  UMM09630-46-2U   
120      97794   5493117  2023-02-08           6        2023  UEU57935-93-6A   
121     

In [3]:
# What three item categories had the most entries?
cat = df['category'].value_counts()
cat.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [4]:
# For the category with the most entries, which subcategory had the most entries?
# Utilize idxmax to return the index of the maximum subcat with the most entries 
# for the category with most entries
cat2 = cat.idxmax()
cat2_df = df[df['category'] == cat2]

sub_cat = cat2_df['subcategory'].value_counts()
highest_cat2 = sub_cat.idxmax()
highest_cat2

'bathroom supplies'

In [5]:
# Which five clients had the most entries in the data?
client_count = df['client_id'].value_counts().head(5)
client_count


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [41]:
# Store the client ids of those top 5 clients in a list.
# retain the index of the 5 clients count df
# output the list of the client ids of the top 5 clients in a python list
client_count.index.to_list()


[33615, 66037, 46820, 38378, 24741]

In [46]:
# How many total units (the qty column) 
# did the client with the most entries order order?
#Use idxmax to get the count of the total units of the client w most entries
#retrieve the highest client with most entries from the client ID
#estimate/sum the count of order based on the quantity ordered by client
highest_client_id = df['client_id'].value_counts().idxmax()
highest_client = df[df['client_id'] == highest_client_id]
sum_of_hghest_qty = highest_client['qty'].sum()
sum_of_hghest_qty

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [55]:
# Create a column that calculates the subtotal 
# for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
#Show first two rows with three columns of client 
# with the most highest purchase based on order quantity
# Note: Include - unit_price, qty, and line_subtotal
df[['unit_price', 'qty', 'line_subtotal']].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [6]:
# Create a column for shipping price.
df['total_weight'] = df['unit_weight'] * df['qty']
# Assume a shipping price of $7 per pound for orders over 50 pounds 
# and $10 per pound for items 50 pounds or under.
#think of lambda + expression (limited expression)
df['shipping_price'] = df['total_weight'].apply(lambda a: 7*a 
                                                if a > 50 
                                                else 10*a)
#show first 3 customers with the highest purchase 
#for unit_price, unit_weight, qty, total_weigh, shipping_price 
df[['unit_price','unit_weight','qty','total_weight','shipping_price']].head(3)



unit_price  unit_weight  qty  total_weight  shipping_price
0     1096.80         7.50  105        787.50         5512.50
1       24.95         1.49   21         31.29          312.90
2       13.52         1.68   39         65.52          458.64

In [7]:
# Create a column for the total price using the subtotal 
# and the shipping price along with a sales tax of 9.25%
df['line_subtotal']= df['unit_price'] *df['qty']
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda a: 7 * a 
                                                if a >50
                                                else 10 * a)
df['line_price'] = df['line_subtotal'] + df['shipping_price']
df['line_price'] *= 1.0925
#show line subtotal: Shipping_price, line_price for first 3 rows
df[['line_subtotal','shipping_price','line_price']].head(3).round(2)

line_subtotal  shipping_price  line_price
0      115164.00         5512.50   131839.08
1         523.95          312.90      914.26
2         527.28          458.64     1077.12

In [77]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly 
# what is charged to the client).
df['line_subtotal'] = df['unit_price']*df['qty']
df['total_weight'] = df['unit_weight']*df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda a: 7*a
                                                if a > 50
                                                else 10 *a)
df['line_cost'] = (df['unit_cost']*df['qty'])+df['shipping_price']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ...  qty line_number line_subtotal  shipping_calc  \
0          17        2023  ...  105           1     115164.00         787.50   
1          20        2023  ...   21           0        523.95          31.29   
2           4        2023  ...   39           6        527.28          65.52   

   shipping_$  total_weight  shipping_price   total_price     line_price  \
0     5512.50        787.50         5512.50  11162.576250  131839.076250   
1      312.90         31.29          312.90     77.408625     914.258625   
2      458.64         65.52          458.64     91.197600    1077.117600   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  

[3 rows x 26 columns]

In [4]:
# Create a column for profit of each line using line cost and line price
df['line_subtotal'] = df['unit_price'] * df['qty']
df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df['total_weight'].apply(lambda z:7*z
                                                if z > 50 
                                                else 10*z)
df['line_cost'] = (df['unit_cost'])*df['qty'] + df['shipping_price']
df['line_price'] = (df['line_subtotal'] +df['shipping_price'])
df['line_price'] *= 1.0925
df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)




first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price  line_cost     line_price  \
0      115164.00        787.50         5512.50   85597.05  131839.076250   
1         523.95         31.29          312.90     629.79     914.258625   
2         527.28         65.52          458.64     765.18    1077.117600   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above from the pd.read_csv client_Dataset.csv
#print total order for order_id - 2742071, 2173913, 6128929 showing total_price
#show output such as Order 2742071 total: $152811.89
orders = ['2742071','2173913','6128929']
for order_id in orders:
    set_order = df[df['order_id'] == int(2742071)]
    set_order2 = df[df['order_id'] == int(2173913)]
    set_order3 = df[df['order_id'] == int(6128929)]
    order_price = set_order['line_price'].sum()
    order_price2 = set_order2['line_price'].sum()
    order_price3 = set_order3['line_price'].sum()

print(f"Order {order_id} total: ${order_price:.2f}")
print(f"Order {order_id} total: ${order_price2:.2f}")
print(f"Order {order_id} total: ${order_price3:.2f}")



Order 6128929 total: $152811.90
Order 6128929 total: $162388.72
Order 6128929 total: $923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [22]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
t5_clients = df.groupby('client_id')['line_price'].sum().nlargest(5)
client_spending = {}
for client_id in t5_clients.index:
    client_data = df[df['client_id'] == client_id]
    total_spent = client_data['line_price'].sum()
    client_spending[client_id] = total_spent
for client_id, total_spent in client_spending.items():
    print(f"{client_id} : ${total_spent:.2f}")


78965 : $655923670.06
85931 : $237417001.42
38473 : $183552571.81
86721 : $181266297.07
84344 : $106648111.60


In [43]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
sum_df = df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('line_price', 'sum'),
    total_cost=('line_cost', 'sum'),
    total_profit=('line_profit', 'sum')
)

t5_clients = sum_df.nlargest(5, 'total_units_purchased')

t5_clients.head(5)

total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
78965                    4033302            5340756.88   6.559237e+08   
86721                    1010426           64335727.72   1.812663e+08   
30143                     846069            8981905.24   3.332885e+07   
85931                     644423           10122549.19   2.374170e+08   
77354                     506382           23814685.86   7.194265e+07   

             total_cost  total_profit  
client_id                              
78965      4.085708e+08  2.473528e+08  
86721      1.382916e+08  4.297470e+07  
30143      2.207846e+07  1.125039e+07  
85931      1.281229e+08  1.092941e+08  
77354      5.951039e+07  1.243226e+07

In [53]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 

# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 
t_5_clients = sum_df.nlargest(5, 'total_units_purchased')


money_columns = ['total_shipping_price', 'total_revenue', 'total_profit', 'total_cost']  # Add 'total_cost' if it exists


def currency_format_millions(val):
    return val / 1_000_000


for mon_col in money_columns:
    t_5_clients[mon_col] = t_5_clients[mon_col].map(currency_format_millions)


t_5_clients.rename(columns={
    'total_units_purchased': 'Units',
    'total_shipping_price': 'Shipping (millions)',
    'total_revenue': 'Total Revenue (millions)',
    'total_profit': 'Total Profit (millions)',
    'total_cost': 'Total Cost (millions)'  
}, inplace=True)


t_5_clients.head(5)

Units  Shipping (millions)  Total Revenue (millions)  \
client_id                                                           
78965      4033302             5.340757                655.923670   
86721      1010426            64.335728                181.266297   
30143       846069             8.981905                 33.328850   
85931       644423            10.122549                237.417001   
77354       506382            23.814686                 71.942652   

           Total Cost (millions)  Total Profit (millions)  
client_id                                                  
78965                 408.570828               247.352842  
86721                 138.291602                42.974695  
30143                  22.078464                11.250386  
85931                 128.122873               109.294129  
77354                  59.510392                12.432260

In [61]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sort_t_5_clients = t_5_clients.sort_values(by='Total Profit (millions)', ascending=False)
sort_t_5_clients.head(5)

Units  Shipping (millions)  Total Revenue (millions)  \
client_id                                                           
78965      4033302             5.340757                655.923670   
85931       644423            10.122549                237.417001   
86721      1010426            64.335728                181.266297   
77354       506382            23.814686                 71.942652   
30143       846069             8.981905                 33.328850   

           Total Cost (millions)  Total Profit (millions)  
client_id                                                  
78965                 408.570828               247.352842  
85931                 128.122873               109.294129  
86721                 138.291602                42.974695  
77354                  59.510392                12.432260  
30143                  22.078464                11.250386

In [8]:
#Write a brief 2-3 sentence summary of your findings
"""
Note: Summary is also included in the README.md file
PART ONE:
On part one we sort the data provided for client with the highest purchase
total quantity purchased of clients with the highest purchase from client ID 

PART TWO:
We surf the data to look for client that spent the most money
   
PART THREE:
We validated from the email reciept of client that spent the most money

PART FOUR:
We also used the validated information to find additional information 
of the client that spent the most money
   """

'\nPART ONE:\nOn part one we sort the data provided for client with the highest purchase\ntotal quantity purchased of clients with the highest purchase from client ID \n\nPART TWO:\nWe surf the data to look for client that spent the most money\n   \nPART THREE:\nWe validated from the email reciept of client that spent the most money\n\nPART FOUR:\nWe also used the validated information to find additional information \nof the client that spent the most money\n   '